Rework ln invoice tags

Christewart opened the pull request:
This PR aims to rework the tag portion of the LN invoices. This is mostly restructing at this point, and moving serialization logic out of the abstract class `LnInvoiceTags` into the actual individual tags themselves.

Christewart committed:
Part 1: Breaking out Bech32 specific functions into a util class, don't embed in Bech32Address

rework ln invoices tags

fix more method names in Bech32

Rename ScriptPubKeyTag -> NodeIdTag

Christewart commented on core/src/main/scala/org/bitcoins/core/util/Bech32.scala:
```diff
@@ -110,27 +113,60 @@ sealed abstract class Bech32 {
 
   /** Converts a byte vector from base 8 to base 5 */
-  def fromBase8ToBase5(bytes: ByteVector): Try[Vector[UInt8]] = {
+  def fromBase8ToBase5(bytes: ByteVector): Vector[UInt8] = {
```
note all of these names are misleding, it isn't really base8. What it really should be named to is something like 

`byteTo5bit`

Christewart commented:
NOTE: Unit tests do not pass and the serialization scheme is still wrong on this PR.

TannerR1776 commented on core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala:
```diff
@@ -1,98 +1,129 @@
-      val bech32DataLength = bech32EncodeDataLength(bech32String)
-      prefix.value + bech32DataLength + bech32String
-    } else { "" }
+  case class ScriptPubKeyTag(spk: ScriptPubKey) extends LnInvoiceTag {
```
This is meant to be the `n` tag which denotes the "33-byte public key of the payee node." I think you are getting it confused with the `f` tag which denotes the "Fallback on-chain address: for bitcoin, this starts with a 5-bit version and contains a witness program or P2PKH or P2SH address." ?

TannerR1776 commented on core/src/main/scala/org/bitcoins/core/protocol/ln/LnTagPrefix.scala:
```diff
@@ -19,7 +19,7 @@ object LnTagPrefix {
   }
 
-  case object SignaturePubKey extends LnTagPrefix {
+  case object ScriptPubKey extends LnTagPrefix {
```
Again, this is for tag `n` which denotes the "33-byte public key of the payee node". The name SignaturePubKey probably comes from this line in BOLT-11: "A writer MAY include one n field, which MUST be set to the public key used to create the signature." We could probably rename to something more intuitive?

TannerR1776 commented on core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala:
```diff
@@ -1,98 +1,129 @@
+  }
 
-  def descriptionHash: Option[Sha256Digest]
+  override def toString: String = {
```
I like splitting out the Tags into their own objects, and giving each their own serialization / `toString` capabilities but, as far as I can tell, none of the encoding used in the `LnInvoiceTag` objects is correct. Unfortunately, it isn't as simple as just converting the data to bytes and Bech32 encoding it. We will need to go back and revise.

TannerR1776 reviewed (commented):
I have reviewed the code. Overall, most changes are good and tests are passing. I left just a few questions for more detail, but I am fine with merging as is. I did note, we will need to go back and re-add serialization for `LnInvoice`'s

Christewart committed:
All invoice tags tests passing except weird serialization order one

Address code review, add some more comments

rename 'LnInvoiceTags' -> 'LnInvoiceTaggedFields'

TannerR1776 commented on core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala:
```diff
@@ -105,23 +174,23 @@ sealed abstract class LnInvoiceTags {
+    b.append(fallbackAddress.map(_.toString).getOrElse(empty))
 
-    //TODO: add routing info
+    val routes = routingInfo.map(_.toString).getOrElse(empty)
```
Is there a reason these two lines aren't just combined into `b.append(routingInfo.map(_.toString).getOrElse(empty))`?

Christewart replied:
It gets a little long. I like using intermediate `vals` if the thing we are passing to a method is not trivial.

TannerR1776 commented on core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala:
```diff
@@ -138,7 +207,7 @@ sealed abstract class LnInvoiceTags {
     val remainder = Bech32EncodedString.length % 32
     val v = Vector(UInt8(quotient.toShort), UInt8(remainder.toShort))
-    Bech32.encode8bitToString(v)
+    Bech32.encode5bitToString(v)
```
This function is orphaned, and should be removed. Functionality has been moved to `dataLength` and `encodeNumber`.

TannerR1776 commented on core/src/main/scala/org/bitcoins/core/protocol/ln/LnInvoiceTags.scala:
```diff
@@ -66,23 +96,62 @@ object LnInvoiceTag {
 
-    override val bytes: ByteVector = u64.bytes
+    override val bytes: ByteVector = {
+      encodeNumber(u64.toInt)
```
This isnt safe unless we add some `require` statements to the the` CtlvExpiryTag` and `ExpiryTimeTag`.

TannerR1776 reviewed (commented):
A lot of changes to `LnInvoiceTags`, but the tests are passing and the serialization looks good. I left a few comments.

TannerR1776 commented:
Awesome stuff!
